ccl/multiregionccl: TestRegionLivenessProberForLeases failed

ccl/multiregionccl.TestRegionLivenessProberForLeases [failed](https://tanzanite.cluster.engflow.com/invocations/default/fcf3cfe4-e213-4b7c-9ace-6a1835bb9981?testReportRun=2&testReportShard=12&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on master @ [1d2534b831775dd84edcf7372d38e35ed4903278](https://github.com/cockroachdb/cockroach/commits/1d2534b831775dd84edcf7372d38e35ed4903278):


```
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/count.go:133
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/regionliveness.LiveRegions.ForEach
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/regionliveness/prober.go:68
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.countLeasesByRegion
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/count.go:120
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.CountLeases.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/count.go:68
        	            	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1736
        	            	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func4
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1823
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:1032
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.runTxn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1079
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1042
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1017
        	            	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1810
        	            	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1737
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.CountLeases
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/count.go:55
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).WaitForNoVersion
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/lease.go:137
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/multiregionccl_test.TestRegionLivenessProberForLeases
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/multiregionccl_test/pkg/ccl/multiregionccl/regionliveness_test.go:353
        	            	  | [...repeated from below...]
        	            	Wraps: (7) failed to read query result
        	            	Wraps: (8) candidate pg code: 57014
        	            	Wraps: (9) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/util/cancelchecker.init
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/cancelchecker/cancel_checker.go:80
        	            	  | runtime.doInit1
        	            	  | 	GOROOT/src/runtime/proc.go:6757
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6724
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:249
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1650
        	            	Wraps: (10) query execution canceled
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *timeutil.TimeoutError (4) *withstack.withStack (5) *errutil.withPrefix (6) *withstack.withStack (7) *errutil.withPrefix (8) *pgerror.withCandidateCode (9) *withstack.withStack (10) *errutil.leafError
        	Test:       	TestRegionLivenessProberForLeases
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestRegionLivenessProberForLeases1792850790
--- FAIL: TestRegionLivenessProberForLeases (31.47s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=2</code>
 - <code>shard=12</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegionLivenessProberForLeases.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-35749

ccl/multiregionccl.TestRegionLivenessProberForLeases [failed](https://tanzanite.cluster.engflow.com/invocations/default/55fd254f-eb71-4570-b9ce-059f0736f1a4?testReportRun=6&testReportShard=12&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on master @ [03525787deade5395bac0a8b5735fa7e715e22ef](https://github.com/cockroachdb/cockroach/commits/03525787deade5395bac0a8b5735fa7e715e22ef):


```
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/count.go:161
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.CountLeases.func1.1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/count.go:97
        	            	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).WithSyntheticDescriptors
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:162
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.CountLeases.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/count.go:91
        	            	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1739
        	            	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func4
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1826
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:1032
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.runTxn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1079
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1042
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1017
        	            	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1813
        	            	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1740
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.CountLeases
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/count.go:73
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).WaitForNoVersion
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/lease.go:141
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/multiregionccl_test.TestRegionLivenessProberForLeases
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/multiregionccl_test/pkg/ccl/multiregionccl/regionliveness_test.go:353
        	            	  | [...repeated from below...]
        	            	Wraps: (7) failed to read query result
        	            	Wraps: (8) candidate pg code: 57014
        	            	Wraps: (9) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/util/cancelchecker.init
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/cancelchecker/cancel_checker.go:80
        	            	  | runtime.doInit1
        	            	  | 	GOROOT/src/runtime/proc.go:6757
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6724
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:249
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1650
        	            	Wraps: (10) query execution canceled
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *timeutil.TimeoutError (4) *withstack.withStack (5) *errutil.withPrefix (6) *withstack.withStack (7) *errutil.withPrefix (8) *pgerror.withCandidateCode (9) *withstack.withStack (10) *errutil.leafError
        	Test:       	TestRegionLivenessProberForLeases
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestRegionLivenessProberForLeases3915310910
--- FAIL: TestRegionLivenessProberForLeases (31.71s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=12</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegionLivenessProberForLeases.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/multiregionccl/BUILD.bazel
+++ b/pkg/ccl/multiregionccl/BUILD.bazel
@@ -86,7 +86,6 @@ go_test(
         "//pkg/sql/sem/tree",
         "//pkg/sql/sqlinstance/instancestorage",
         "//pkg/sql/sqlliveness",
-        "//pkg/sql/sqlliveness/slbase",
         "//pkg/sql/sqlliveness/slstorage",
         "//pkg/sql/sqltestutils",
         "//pkg/testutils",

--- a/pkg/ccl/multiregionccl/regionliveness_test.go
+++ b/pkg/ccl/multiregionccl/regionliveness_test.go
@@ -32,7 +32,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/regionliveness"
 	"github.com/cockroachdb/cockroach/pkg/sql/regions"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage"
-	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slbase"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
@@ -44,7 +43,14 @@ import (
 )
 
 const (
+	// Use a low probe timeout number, but intentionally only manipulate to
+	// this value when *failures* are expected.
 	testingRegionLivenessProbeTimeout = time.Second * 2
+	// Use a long probe timeout by default when running this test (since CI
+	// can easily hit query timeouts).
+	testingRegionLivenessProbeTimeoutLong = time.Minute
+	// Use a reduced liveness TTL for region liveness infrastructure.
+	testingRegionLivenessTTL = time.Second * 10
 )
 
 func TestRegionLivenessProber(t *testing.T) {
@@ -53,7 +59,7 @@ func TestRegionLivenessProber(t *testing.T) {
 	// This test forces the SQL liveness TTL be a small number,
 	// which makes the heartbeats even more critical. Under stress and
 	// race environments this test becomes even more sensitive, if
-	// we can't send heartbeats within 10 seconds.
+	// we can't send heartbeats within 20 seconds.
 	skip.UnderStress(t)
 	skip.UnderRace(t)
 	skip.UnderDeadlock(t)
@@ -65,10 +71,10 @@ func TestRegionLivenessProber(t *testing.T) {
 	makeSettings := func() *cluster.Settings {
 		cs := cluster.MakeTestingClusterSettings()
 		instancestorage.ReclaimLoopInterval.Override(ctx, &cs.SV, 150*time.Millisecond)
-		slbase.DefaultTTL.Override(ctx, &cs.SV, 10*time.Second)
 		regionliveness.RegionLivenessEnabled.Override(ctx, &cs.SV, true)
 		return cs
 	}
+	defer regionliveness.TestingSetUnavailableAtTTLOverride(testingRegionLivenessTTL)()
 
 	expectedRegions := []string{
 		"us-east",
@@ -88,11 +94,11 @@ func TestRegionLivenessProber(t *testing.T) {
 	var tenants []serverutils.ApplicationLayerInterface
 	var tenantSQL []*gosql.DB
 	blockProbeQuery := atomic.Bool{}
-	defer regionliveness.TestingSetProbeLivenessTimeout(500*time.Millisecond,
+	defer regionliveness.TestingSetProbeLivenessTimeout(
 		func() {
 			// Timeout attempts to probe intentionally.
 			if blockProbeQuery.Swap(false) {
-				time.Sleep(2 * time.Second)
+				time.Sleep(testingRegionLivenessProbeTimeout)
 			}
 		})()
 
@@ -148,6 +154,11 @@ func TestRegionLivenessProber(t *testing.T) {
 	liveRegions, err = regionProber.QueryLiveness(ctx, testTxn)
 	require.NoError(t, err)
 	checkExpectedRegions(expectedRegions, liveRegions)
+	// Override the table timeout probe for testing to ensure timeout failures
+	// happen now.
+	for _, ts := range tenants {
+		regionliveness.RegionLivenessProbeTimeout.Override(ctx, &ts.ClusterSettings().SV, testingRegionLivenessProbeTimeout)
+	}
 	// Probe the liveness of the region, but timeout the query
 	// intentionally to make it seem dead.
 	blockProbeQuery.Store(true)
@@ -213,7 +224,7 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 	// This test forces the SQL liveness TTL be a small number,
 	// which makes the heartbeats even more critical. Under stress and
 	// race environments this test becomes even more sensitive, if
-	// we can't send heartbeats within 10 seconds.
+	// we can't send heartbeats within 20 seconds.
 	skip.UnderStress(t)
 	skip.UnderRace(t)
 	skip.UnderDeadlock(t)
@@ -225,10 +236,10 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 	makeSettings := func() *cluster.Settings {
 		cs := cluster.MakeTestingClusterSettings()
 		instancestorage.ReclaimLoopInterval.Override(ctx, &cs.SV, 150*time.Millisecond)
-		slbase.DefaultTTL.Override(ctx, &cs.SV, 10*time.Second)
 		regionliveness.RegionLivenessEnabled.Override(ctx, &cs.SV, true)
 		return cs
 	}
+	defer regionliveness.TestingSetUnavailableAtTTLOverride(testingRegionLivenessTTL)()
 
 	expectedRegions := []string{
 		"us-east",
@@ -239,11 +250,11 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 	targetCount := atomic.Int64{}
 	var tenants []serverutils.ApplicationLayerInterface
 	var tenantSQL []*gosql.DB
-	defer regionliveness.TestingSetProbeLivenessTimeout(1*time.Second, func() {
+	defer regionliveness.TestingSetProbeLivenessTimeout(func() {
 		if !detectLeaseWait.Load() {
 			return
 		}
-		time.Sleep(time.Second * 2)
+		time.Sleep(testingRegionLivenessProbeTimeout)
 		targetCount.Swap(0)
 		detectLeaseWait.Swap(false)
 	})()
@@ -274,7 +285,7 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 							if targetCount.Add(1) != 1 {
 								return
 							}
-							time.Sleep(time.Second * 2)
+							time.Sleep(testingRegionLivenessProbeTimeout)
 						}
 					},
 				},
@@ -298,10 +309,7 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 			require.NoError(t, err)
 		}
 	}
-	// Override the table timeout probe for testing.
-	for _, ts := range tenants {
-		regionliveness.RegionLivenessProbeTimeout.Override(ctx, &ts.ClusterSettings().SV, testingRegionLivenessProbeTimeout)
-	}
+
 	// Create a new table and have it used on all nodes.
 	_, err = tenantSQL[0].Exec("CREATE TABLE t1(j int)")
 	require.NoError(t, err)
@@ -312,6 +320,11 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 	row := tenantSQL[0].QueryRow("SELECT 't1'::REGCLASS::OID")
 	var tableID int
 	require.NoError(t, row.Scan(&tableID))
+	// Override the table timeout probe for testing to ensure timeout failures
+	// happen now.
+	for _, ts := range tenants {
+		regionliveness.RegionLivenessProbeTimeout.Override(ctx, &ts.ClusterSettings().SV, testingRegionLivenessProbeTimeout)
+	}
 	// Issue a schema change which should mark this region as dead, and fail
 	// out because our probe query will time out.
 	detectLeaseWait.Swap(true)
@@ -350,5 +363,9 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 		})
 		return builder.BuildExistingMutableType()
 	})
+	// Restore the override to reduce the risk of failing on overloaded systems.
+	for _, ts := range tenants {
+		regionliveness.RegionLivenessProbeTimeout.Override(ctx, &ts.ClusterSettings().SV, testingRegionLivenessProbeTimeoutLong)
+	}
 	require.NoError(t, lm.WaitForNoVersion(ctx, descpb.ID(tableID), cachedDatabaseRegions, retry.Options{}))
 }

--- a/pkg/sql/regionliveness/prober.go
+++ b/pkg/sql/regionliveness/prober.go
@@ -122,16 +122,21 @@ type livenessProber struct {
 	settings        *clustersettings.Settings
 }
 
-var probeLivenessTimeout = 15 * time.Second
 var testingProbeQueryCallbackFunc func()
+var testingUnavailableAtTTLOverride time.Duration
 
-func TestingSetProbeLivenessTimeout(newTimeout time.Duration, probeCallbackFn func()) func() {
-	oldTimeout := probeLivenessTimeout
-	probeLivenessTimeout = newTimeout
+func TestingSetProbeLivenessTimeout(probeCallbackFn func()) func() {
 	testingProbeQueryCallbackFunc = probeCallbackFn
 	return func() {
-		probeLivenessTimeout = oldTimeout
 		probeCallbackFn = nil
+	}
+}
+
+func TestingSetUnavailableAtTTLOverride(duration time.Duration) func() {
+	oldValue := testingUnavailableAtTTLOverride
+	testingUnavailableAtTTLOverride = duration
+	return func() {
+		testingUnavailableAtTTLOverride = oldValue
 	}
 }
 
@@ -211,6 +216,9 @@ func (l *livenessProber) ProbeLivenessWithPhysicalRegion(
 	// Region has gone down, set the unavailable_at time on it
 	return l.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
 		defaultTTL := slbase.DefaultTTL.Get(&l.settings.SV)
+		if testingUnavailableAtTTLOverride != 0 {
+			defaultTTL = testingUnavailableAtTTLOverride
+		}
 		defaultHeartbeat := slbase.DefaultHeartBeat.Get(&l.settings.SV)
 		// Get the read timestamp and pick a commit deadline.
 		commitDeadline := txn.ReadTimestamp().AddDuration(defaultHeartbeat)
@@ -316,7 +324,8 @@ func (l *livenessProber) GetProbeTimeout() (bool, time.Duration) {
 // when checking for region liveness.
 func IsQueryTimeoutErr(err error) bool {
 	return pgerror.GetPGCode(err) == pgcode.QueryCanceled ||
-		errors.HasType(err, (*timeutil.TimeoutError)(nil))
+		errors.HasType(err, (*timeutil.TimeoutError)(nil)) ||
+		errors.Is(err, context.DeadlineExceeded)
 }
 
 // IsMissingRegionEnumErr determines if a query hit an error because of a missing

